From e04c948c2075930387e258e4d3e3068302544768 Mon Sep 17 00:00:00 2001
From: Antonio Terceiro <terceiro@debian.org>
Date: Tue, 5 Oct 2021 20:27:11 -0300
Subject: [PATCH 1/2] Fix indexing with compressed APT Packages files

APT can be configured to keep its Packages files compressed after
downloading, and these days this is even a common setup. Therefore, one
cannot assume that *Packages files under /var/lib/apt/lists/ exist, or
can be read as plain text.

Instead, use the documented APT interfaces for listing and getting the
contents of the Packages files, that work whether they are compressed or
not.

Closes: #617856
---
 apt-show-versions | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/apt-show-versions b/apt-show-versions
index 9bfa3b0..240ca7c 100755
--- a/apt-show-versions
+++ b/apt-show-versions
@@ -193,7 +193,7 @@ if (-e $filescachefile and -M $filescachefile < -M $list_dir) {
 # Test also to be sure $filescachefile is not corrupt and returns a ref to it
 if (!-e $filescachefile or -M $list_dir < -M $filescachefile or !ref($filesref)) {
     opendir(DIR, $list_dir) or &die("Can't opendir $list_dir: $!\n");
-    @files = map { $list_dir . $_} grep /Packages$/, readdir(DIR);
+    @files = split(/\s/, `apt-get indextargets -o Dir::State::lists=${list_dir} --format='\$(FILENAME)' 'Created-By: Packages'`);
     ($< == 0) and (nstore(\@files, $filescachefile) or
         warn "Can't write $filescachefile\n");
     ($< == 0) and (chmod($cachefilemode, $filescachefile) or
@@ -569,7 +569,7 @@ sub parse_file {
     my ($key, $value, $package, $packages);
 
     my $release = &determine_pkgfile_release($file);
-    open FILE, $file or &die("Can't open file $file: $!\n");
+    open FILE, "/usr/lib/apt/apt-helper cat-file $file|" or &die("Can't open file $file: $!\n");
     if ($opts{'verbose'}) {print "Parsing $file...";};
     while (<FILE>) {
         if (/^$/){
-- 
2.33.0

